Group additions to Builder order (#233)
buildpacks/spec: extensions/project-descriptor.md

The `api` and `inline` key MUST be defined in the `script` table. The value of the `inline` key will be used as the build script for the [inline buildpack](#Definitions) this entry represents. The value of the `api` key defines its Buildpack API compatibility, and the `shell` key defines the shell used to execute the `inline` script.

#### `io.buildpacks.pre.group` (optional)

> This table MAY contain a list of buildpacks to insert at the beginning of an automatically detected group. Given an order with multiple groups, the list of `pre` buildpacks will be inserted at the beginning of each group such that the are run as if they were original included in the group. Each phase of the injected buildpack(s) will execute as normal.

Reviewer: ```suggestion
This table MAY contain a list of buildpacks to insert at the beginning of an automatically detected group. Given an order with multiple groups, the list of `pre` buildpacks will be inserted at the beginning of each group such that they are run as if they were original included in the group. Each phase of the injected buildpack(s) will execute as normal.
```

> This table MAY contain a list of buildpacks to insert at the beginning of an automatically detected group. Given an order with multiple groups, the list of `pre` buildpacks will be inserted at the beginning of each group such that they are run as if they were original included in the group. Each phase of the injected buildpack(s) will execute as normal.

Reviewer 2: ```suggestion
This table MAY contain a list of buildpacks to insert at the beginning of an automatically detected group. Given an order with multiple groups, the list of `pre` buildpacks will be inserted at the beginning of each group such that they are run as if they were originally included in the group. Each phase of the injected buildpack(s) will execute as normal.
```

Reviewer 3: The RFC states:
> injected into the beginning or end of the groups defined in an order.toml provided by a builder.

I initially read `insert at the beginning of an automatically detected group` as a post-detection step. But I think it is meant to say something like `insert at the beginning of all defined groups`.

Reviewer 3: Another question regarding insertion of these buildpacks. Do these insertions happen to the group defined _inside_ of a project descriptor? Or do these only get injected for builder's `order.toml`?

Author: @jabrown85 

>But I think it is meant to say something like insert at the beginning of all defined groups.

oh, that's a subtle but important different. I think it means detect will be run. I'll update this PR to use that langauge.

> Do these insertions happen to the group defined inside of a project descriptor? Or do these only get injected for builder's order.toml?

Yes, i'll add that.

Author: Actually, i take that last part back. There shouldn't be a need to use `pre/`post` if you're defining a group in project.toml. I'm not sure what the behavior should be if both are used though.

Reviewer 4: > Actually, i take that last part back. There shouldn't be a need to use `pre/`post` if you're defining a group in project.toml. I'm not sure what the behavior should be if both are used though.

maybe it should be an "or", where you can't use `pre`/`post` but not with a normal buildpack group.

Reviewer 4: From the leadership WG meeting, what happens on the `pack` CLI when overriding with `-b`? Will builders support `pre`/`post` with considerations like buildpacks/rfcs#175?

> This table MAY contain a list of buildpacks to insert at the beginning of an automatically detected group. Given an order with multiple groups, the list of `pre` buildpacks will be inserted at the beginning of each group such that they are run as if they were originally included in the group. Each phase of the injected buildpack(s) will execute as normal.

Author: ```suggestion
This table MAY contain a list of buildpacks to insert at the beginning of an automatically detected group. Given an order with multiple groups, the list of `pre` buildpacks will be inserted at the beginning of each automatically detected group such that they are run as if they were originally included in the group. Each phase of the injected buildpack(s) will execute as normal.
```

> This table MAY contain a list of buildpacks to insert at the beginning of an automatically detected group. Given an order with multiple groups, the list of `pre` buildpacks will be inserted at the beginning of each automatically detected group such that they are run as if they were originally included in the group. Each phase of the injected buildpack(s) will execute as normal.

Reviewer: I'm curious about the use of the term "detected" here. It's my understanding that pre and post buildpacks would _always_ execute for every declared group. This phrasing makes it sounds like the pre and post buildpacks would only execute after the initial declared groups have gone through detection and a certain group was selected.

Reviewer: This seems to be a replay of [this comment](https://github.com/buildpacks/spec/pull/233#discussion_r662515350) which doesn't seem to have been addressed.

Reviewer 4: @jromero does this make it clearer?

```suggestion
This table MAY contain a list of buildpacks to insert at the beginning of every group. Given an order with multiple groups, the list of `pre` buildpacks will be inserted at the beginning of each group such that they are run as if they were originally included in the group. Each phase of the injected buildpack(s) will execute as normal.
```

Reviewer 5: I agree with @jromero 's concern and prefer @hone 's suggestion to the current wording.

> [[build.post.buildpacks]]

Reviewer: ```suggestion
```

This seems to be out of place.

The schema for this table is identical to [`io.buildpacks.group`](#iobuildpacksgroup-optional)

#### `io.buildpacks.post.group` (optional)

> This table MAY contain a list of buildpacks to insert at the end of an automatically detected group. Given an order with multiple groups, the list of `post` buildpacks will be inserted at the end of each group such that the are run as if they were original included in the group. Each phase of the injected buildpack(s) will execute as normal.

Reviewer: ```suggestion
This table MAY contain a list of buildpacks to insert at the end of an automatically detected group. Given an order with multiple groups, the list of `post` buildpacks will be inserted at the end of each group such that they are run as if they were original included in the group. Each phase of the injected buildpack(s) will execute as normal.
```

> This table MAY contain a list of buildpacks to insert at the end of an automatically detected group. Given an order with multiple groups, the list of `post` buildpacks will be inserted at the end of each group such that they are run as if they were original included in the group. Each phase of the injected buildpack(s) will execute as normal.

Reviewer 2: ```suggestion
This table MAY contain a list of buildpacks to insert at the end of an automatically detected group. Given an order with multiple groups, the list of `post` buildpacks will be inserted at the end of each group such that they are run as if they were originally included in the group. Each phase of the injected buildpack(s) will execute as normal.
```

> This table MAY contain a list of buildpacks to insert at the end of an automatically detected group. Given an order with multiple groups, the list of `post` buildpacks will be inserted at the end of each group such that they are run as if they were originally included in the group. Each phase of the injected buildpack(s) will execute as normal.

Author: ```suggestion
This table MAY contain a list of buildpacks to insert at the end of an automatically detected group. Given an order with multiple groups, the list of `post` buildpacks will be inserted at the end of each automatically detected group such that they are run as if they were originally included in the group. Each phase of the injected buildpack(s) will execute as normal.
```

> This table MAY contain a list of buildpacks to insert at the end of an automatically detected group. Given an order with multiple groups, the list of `post` buildpacks will be inserted at the end of each automatically detected group such that they are run as if they were originally included in the group. Each phase of the injected buildpack(s) will execute as normal.

Reviewer: Same concerns about "detected" term usages here as above.

Author: ```suggestion
This table MAY contain a list of buildpacks to insert at the end of every group. Given an order with multiple groups, the list of `post` buildpacks will be inserted at the end of each group such that they are run as if they were originally included in the group. Each phase of the injected buildpack(s) will execute as normal.
```
